reconcile/cna/assets/asset.py

```diff
@@ -1,38 +1,222 @@
 from __future__ import annotations
 from abc import ABC, abstractmethod
-from dataclasses import dataclass, field
+
+from pydantic.dataclasses import dataclass
+from pydantic.fields import FieldInfo
 from enum import Enum
-from typing import Any, Optional
+from typing import Any, Mapping, Optional, Type
+
+from reconcile.gql_definitions.cna.queries.cna_resources import CNAssetV1
+
+
+ASSET_TYPE_FIELD = "asset_type"
+ASSET_PARAMETERS_FIELD = "parameters"
 
 
 class AssetError(Exception):
     pass
 
 
-class AssetType(Enum):
+class UnknownAssetTypeError(Exception):
+    pass
+
+
+class AssetType(str, Enum):
     NULL = "null"
     EXAMPLE_AWS_ASSUMEROLE = "example-aws-assumerole"
 
 
+def asset_type_by_id(asset_type_id: str) -> Optional[AssetType]:
+    try:
+        return AssetType(asset_type_id)
+    except ValueError:
+        return None
+
+
+def asset_type_from_raw_asset(raw_assset: Mapping[str, Any]) -> Optional[AssetType]:
+    return asset_type_by_id(raw_assset.get(ASSET_TYPE_FIELD, ""))
+
+
+class AssetTypeVariableType(Enum):
+    STRING = "${string}"
+    NUMBER = "${number}"
+    LIST_STRING = "${list(string)}"
+    LIST_NUMBER = "${list(number)}"
+
+
+@dataclass(frozen=True)
+class AssetTypeVariable:
+    name: str
+    type: AssetTypeVariableType
+    optional: bool = False
+    default: Optional[str] = None
+
+
+@dataclass
+class AssetTypeMetadata:
+    id: AssetType
+    bindable: bool
+    variables: set[AssetTypeVariable]
+
+
 class AssetStatus(Enum):
+    UNKNOWN = None
     READY = "Ready"
     TERMINATED = "Terminated"
     PENDING = "Pending"
     RUNNING = "Running"
 
 
-@dataclass(frozen=True)
+class AssetModelConfig:
+    allow_population_by_field_name = True
+    extra = "forbid"
+
+
+@dataclass(frozen=True, config=AssetModelConfig)
```

Comment:
Does `extra.forbid` work on `@dataclass`? My understanding is that this would only work with BaseModel, because it generates a different __init__ method?

https://github.com/pydantic/pydantic/issues/710#issuecomment-530446094

Reply:
Ah no actually its the other way around. `@pydantic.dataclass` does not support extra args, even if its configured to be allowed - but thats not what we want - so all godd :)